### PR TITLE
visual patcher initial go

### PR DESCRIPTION
**Pull request by gmarinov**
if in doubt, should contain develop, newui-modals and other recent newui changes.

### CODE REVIEW

File: browser/scripts/application.js [hunk 1]
```diff
@@ -173,6 +173,16 @@ Application.prototype.releaseHoverSlot = function() {
 	this.releaseHoverConnections();
 }
 
+Application.prototype.setSlotCssClasses = function(slot, slot_div) {	/* @var slot_div jQuery */
+	if (typeof slot_div == 'undefined') return false;
```
Comment (kschzt) [inline browser/scripts/application.js]: please install jshint :) it'd tell you to use === instead

File: browser/scripts/application.js [hunk 2]
```diff
@@ -173,6 +173,16 @@ Application.prototype.releaseHoverSlot = function() {
...
+	if (!slot_div) return false;
+	if (!(slot_div === this.hover_slot_div)) slot_div.removeClass('p_connecting')
```
Comment (kschzt) [inline browser/scripts/application.js]: please split this into two lines, and use `if (slot_div !== this.hover_slot_div)` instead (better readability)

File: browser/scripts/application.js [hunk 3]
```diff
@@ -173,6 +173,16 @@ Application.prototype.releaseHoverSlot = function() {
...
+	slot_div.removeClass('p_compatible')
+		.removeClass('p_incompatible');
+	(slot && slot.is_connected) ? slot_div.addClass('p_connected') : slot_div.removeClass('p_connected');
```
Comment (kschzt) [inline browser/scripts/application.js]: please use an if / else for this

File: browser/scripts/application.js [hunk 4]
```diff
@@ -141,7 +142,7 @@ Application.prototype.activateHoverSlot = function() {
 	if(!hs)
 		return;
 
-	this.hover_slot_div[0].style.backgroundColor = E2.erase_color;
+//	this.hover_slot_div[0].style.backgroundColor = E2.erase_color;
```
Comment (kschzt) [inline browser/scripts/application.js]: please remove commented code

File: browser/scripts/application.js [hunk 5]
```diff
@@ -268,16 +281,27 @@ Application.prototype.onMouseReleased = function() {
...
 			c.signal_change(true)
 
 		if (ec.srcSlotDiv)
-			ec.srcSlotDiv[0].style.color = '#000'
+			this.setSlotCssClasses(ec.srcSlot, ec.srcSlotDiv);
 		if (ec.dstSlotDiv)
-			ec.dstSlotDiv[0].style.color = '#000'
+			this.setSlotCssClasses(ec.dstSlot, ec.dstSlotDiv);
+
+		if (success) {
```
Comment (kschzt) [inline browser/scripts/application.js]: please remove console.logs

File: browser/scripts/application.js [hunk 6]
```diff
@@ -475,39 +499,6 @@ Application.prototype.onNodeHeaderClicked = function() {
 
 Application.prototype.onNodeHeaderDblClicked = function(node) {
 
-	var that = this
```
Comment (kschzt) [inline browser/scripts/application.js]: why was this removed again? :)

File: browser/scripts/node-ui.js [hunk 7]
```diff
@@ -1,26 +1,40 @@
-function NodeUI(parent_node, x, y) {
-	var that = this
-	this.parent_node = parent_node;
+UIpoint = function(x,y,z) {
 	this.x = x;
 	this.y = y;
+	this.z = z;
+}
+
+NodeUI = function(parent_node, x, y, z) {
```
Comment (kschzt) [inline browser/scripts/node-ui.js]: why turn this into an anonymous function?

File: browser/scripts/node-ui.js [hunk 8]
```diff
@@ -1,26 +1,40 @@
-function NodeUI(parent_node, x, y) {
-	var that = this
-	this.parent_node = parent_node;
+UIpoint = function(x,y,z) {
```
Comment (kschzt) [inline browser/scripts/node-ui.js]: why turn this into an anonymous function?

File: browser/scripts/node-ui.js [hunk 9]
```diff
@@ -1,26 +1,40 @@
...
+}
+
+NodeUI = function(parent_node, x, y, z) {
+	var that = this
+
+	this.parent_node = parent_node;
+	this.selected = false;
+	this.position = new UIpoint(x,y,z);
+	this.x = x || 0;
+	this.y = y || 0;
```
Comment (kschzt) [inline browser/scripts/node-ui.js]: initialize `z` too, and *then* init `position` with `new UIPoint()`

File: browser/scripts/node-ui.js [hunk 10]
```diff
@@ -36,102 +50,241 @@ function NodeUI(parent_node, x, y) {
...
 	else
 		this.plugin_ui = {}; // We must set a dummy object so plugins can tell why they're being called.
-	
-	make_draggable(dom, 
+
+	if (this.hasPreferences()) {	// create a preferences button and wire it up
+		var inp_config = makeButton(null, 'Edit preferences', 'config_btn')
+		inp_config.removeClass('btn');
+		inp_config.click(function() {
+			// TODO: dispatch event back to parent_node
```
Comment (kschzt) [inline browser/scripts/node-ui.js]: please remove TODO comment :) add an issue on waffle instead

File: browser/scripts/node-ui.js [hunk 11]
```diff
@@ -36,102 +50,241 @@ function NodeUI(parent_node, x, y) {
...
+ */
+NodeUI.prototype.update = function() {
+	if (!this.dom) return;
+	var s = this.dom[0].style;
+
+	var xx = this.position.x;
+	var yy = this.position.y;
+
+	// temp fix for plugins appearing at -98px top
+	// TODO: fix root cause and remove
```
Comment (kschzt) [inline browser/scripts/node-ui.js]: please remove TODO and create a waffle issue instead :)

**Comment by kschzt** [comment]
apart from the minor commits, I love it. this is really nice work! ^_^

File: browser/scripts/node-ui.js (hunk 7 above)
Comment (gmarinov) [inline browser/scripts/node-ui.js]: NodeUI is now a class..

**Comment by kschzt** [comment]
lgtm
